controllers/nstemplatetierrevisioncleanup/nstemplatetier_revision_cleanup_controller.go

```diff
@@ -0,0 +1,141 @@
+	logger := log.FromContext(ctx)
+
+	// fetch the NSTemplateTier tier
+	ttr := &toolchainv1alpha1.TierTemplateRevision{}
+	if err := r.Client.Get(ctx, request.NamespacedName, ttr); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("TierTemplateRevision not found")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("unable to get the current TierTemplateRevision: %w", err)
+	}
+
+	//there is no point in fetching  the NStemplateTier, if the TTR  is just created
+	//get the tier template revision creation time stamp and the duration
+	timeSinceCreation := time.Since(ttr.GetCreationTimestamp().Time)
+
+	//the ttr age should be greater than 30 seconds
+	if timeSinceCreation < DeletionTimeThreshold {
+		requeueAfter := DeletionTimeThreshold - timeSinceCreation
+		logger.Info("the TierTemplateRevision is not old enough", "requeue-after", requeueAfter)
+		return reconcile.Result{RequeueAfter: requeueAfter}, nil
+	}
+
+	//check if there is tier-name label available
+	tierName, ok := ttr.GetLabels()[toolchainv1alpha1.TierLabelKey]
+	if !ok {
+		//if tier-name label is not found we should delete the TTR
+		return reconcile.Result{}, r.deleteTTR(ctx, ttr)
+	}
+	// fetch the related NSTemplateTier tier
+	tier := &toolchainv1alpha1.NSTemplateTier{}
+	if err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: ttr.GetNamespace(),
+		Name:      tierName,
+	}, tier); err != nil {
+		if errors.IsNotFound(err) {
+			// if there is no related nstemplate tier, we can delete the TTR directly
+			return reconcile.Result{}, r.deleteTTR(ctx, ttr)
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("unable to get the current NSTemplateTier: %w", err)
+	}
+
+	// let's make sure that we delete the TTR only if it's not used
+	isTTrUnused, err := r.verifyUnusedTTR(ctx, tier, ttr)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Delete the unused revision
+	if isTTrUnused {
+		return reconcile.Result{}, r.deleteTTR(ctx, ttr)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// verifyUnusedTTR function verifies that the TTR is not used (returns true if it's used).
+// this is done by:
+//   - checking the NSTemplateTier.status.revisions field, if the TTR is referenced there or not
+//   - checking if all Spaces are up-to-date. In case there are some outdated space, we could risk that the TTR is still being used
+func (r *Reconciler) verifyUnusedTTR(ctx context.Context, nsTmplTier *toolchainv1alpha1.NSTemplateTier,
+	rev *toolchainv1alpha1.TierTemplateRevision) (bool, error) {
+
+	logger := log.FromContext(ctx)
+	//check if the ttr name is present status.revisions
+	for _, ttStatusRev := range nsTmplTier.Status.Revisions {
+		if ttStatusRev == rev.Name {
+			logger.Info("the TierTemplateRevision is still being referenced in NSTemplateTier.Status.Revisions")
```

Comment:
Isn't this the expected case? IMHO we should not log this if it is.

Reply:
I believe that this could help us with any future debugging of why some specific (old) TTR is not deleted, so I would keep it there. 
This controller won't get too many reconciles, so it shouldn't populate the logs too much.

Reply:
i would vote for keeping the logger